### PR TITLE
feature(generator): add support for streaming read rpcs

### PR DESCRIPTION
**Pull request by scotthart**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5748)

### CODE REVIEW

File: generator/integration_tests/golden/iam_credentials_client.gcpcxx.pb.cc [hunk 1]
```diff
@@ -62,6 +62,13 @@ IAMCredentialsClient::ListLogs(std::string const& parent) {
   return connection_->ListLogs(request);
 }
 
+TailLogEntriesStream
+IAMCredentialsClient::TailLogEntries(std::vector<std::string> const& resource_names) {
```
Comment (coryan) [inline generator/integration_tests/golden/iam_credentials_client.gcpcxx.pb.cc:66]: The name of these services in the golden files are extremely confusing to me.  Seeing something like `IAMCredentialsClient::TailLogEntries()` raises so many questions.

Consider using a more neutral name, like `GeneratorTestClient`, or `GeneratorGoldenTestsClient`

Comment (scotthart) [reply generator/integration_tests/golden/iam_credentials_client.gcpcxx.pb.cc:66]: created #5810

**Review by coryan — COMMENTED** [review]
> Reviewed 33 of 40 files at r1.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5748)* status: all files reviewed, 2 unresolved discussions (waiting on @scotthart)

___
*[generator/integration_tests/golden/iam_credentials_client.gcpcxx.pb.cc, line 66 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5748#-MSd9BiHMEC7FbeiAHYH-r1-66:-MSi1d3Z0bNXDQOnk2dH:bwh5rdm) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/61119c395fc1151c248994c1472d1326b2f3de60/generator/integration_tests/golden/iam_credentials_client.gcpcxx.pb.cc#L66)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

created #5810 
</blockquote></details>

Thanks.  In the same spirit, you have really complicated request and response parameters in some of these golden files, are those needed?  Maybe you can just create them with a single string field or something?
___
*[generator/integration_tests/golden/iam_credentials_connection.gcpcxx.pb.cc, line 219 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5748#-MSi26aOBIQhl7_n68uv:-MSi26aOBIQhl7_n68uw:b-d1zpqc) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/779e2b7825b000c813fb8d0a4428ddef143ce1e8/generator/integration_tests/golden/iam_credentials_connection.gcpcxx.pb.cc#L219)):*
> ```cpp
>                 std::move(request));
> 
>     if (options_.tracing_enabled("rpc-streams")) {
> ```

I think the stub is already decorated for logging, right?  Can we move this to the `LoggingStub`?
___
*[generator/integration_tests/golden/internal/iam_credentials_logging_decorator.gcpcxx.pb.cc, line 82 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5748#-MSi2qp1B3-EmAyi26KI:-MSi2qp1B3-EmAyi26KJ:btk4dv0) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/779e2b7825b000c813fb8d0a4428ddef143ce1e8/generator/integration_tests/golden/internal/iam_credentials_logging_decorator.gcpcxx.pb.cc#L82)):*
> ```cpp
> }
> 
> std::unique_ptr<grpc::ClientReaderInterface<::google::test::admin::database::v1::TailLogEntriesResponse>>
> ```

Ah, I see, this returns the `grpc::ClientReaderInterface<>`, what if we change the stub to return a `google::cloud::StreamingReadRpc<T>`, can we then push all the logging stuff to this class?

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5748)* status: 17 of 39 files reviewed, 2 unresolved discussions (waiting on @coryan)

___
*[generator/integration_tests/golden/iam_credentials_connection.gcpcxx.pb.cc, line 219 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5748#-MSi26aOBIQhl7_n68uv:-MSjtofF79nj08yOTyIY:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/779e2b7825b000c813fb8d0a4428ddef143ce1e8/generator/integration_tests/golden/iam_credentials_connection.gcpcxx.pb.cc#L219)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I think the stub is already decorated for logging, right?  Can we move this to the `LoggingStub`?
</blockquote></details>

Done.
___
*[generator/integration_tests/golden/internal/iam_credentials_logging_decorator.gcpcxx.pb.cc, line 82 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5748#-MSi2qp1B3-EmAyi26KI:-MSjuUwu3zEi4ZfSZKG7:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/779e2b7825b000c813fb8d0a4428ddef143ce1e8/generator/integration_tests/golden/internal/iam_credentials_logging_decorator.gcpcxx.pb.cc#L82)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Ah, I see, this returns the `grpc::ClientReaderInterface<>`, what if we change the stub to return a `google::cloud::StreamingReadRpc<T>`, can we then push all the logging stuff to this class?
</blockquote></details>

Done.

**Review by coryan — COMMENTED** [review]
Couple of thoughts below.

> Reviewed 22 of 23 files at r3.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5748#-:-MSjxzST5AGIEQmvq9Ce:bbalmad)* status: all files reviewed, 4 unresolved discussions (waiting on @scotthart)

___
*[generator/integration_tests/golden/iam_credentials_connection.gcpcxx.pb.h, line 97 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5748#-MSjvP1Q0IqeE4zbmiaQ:-MSjvP1Q0IqeE4zbmiaR:b-qdi1sv) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/3c51780ee9e5f0db333050bedf4e33776ff27e3e/generator/integration_tests/golden/iam_credentials_connection.gcpcxx.pb.h#L97)):*
> ```objc
> 
> std::shared_ptr<IAMCredentialsConnection> MakeIAMCredentialsConnection(
>     IAMCredentialsConnectionOptions const& options,
> ```

Is this still needed?
___
*[generator/integration_tests/golden/iam_credentials_connection.gcpcxx.pb.cc, line 225 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5748#-MSjvVdW6ZdP6Wf9t8wm:-MSjvVdX747Opgu5Gytx:b-mf6emd) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/3c51780ee9e5f0db333050bedf4e33776ff27e3e/generator/integration_tests/golden/iam_credentials_connection.gcpcxx.pb.cc#L225)):*
> ```cpp
>   std::unique_ptr<BackoffPolicy const> backoff_policy_prototype_;
>   std::unique_ptr<IAMCredentialsConnectionIdempotencyPolicy> idempotency_policy_;
>   IAMCredentialsConnectionOptions options_;
> ```

Is this still needed?  I think you needed it to get the tracing parameters, but no longer use it?
___
*[generator/integration_tests/golden/iam_credentials_metadata_decorator_test.cc, line 172 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5748#-MSjvyVP10fdZ3BQozd-:-MSjvyVP10fdZ3BQozd0:b-hcs7m7) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/3c51780ee9e5f0db333050bedf4e33776ff27e3e/generator/integration_tests/golden/iam_credentials_metadata_decorator_test.cc#L172)):*
> ```cpp
> }
> 
> class MockTailLogEntriesStreamingReadRpc
> ```

Should we write a templated mock and do this just once?  OTOH, it is hard to remember where these little mocks live.
___
*[generator/integration_tests/golden/internal/iam_credentials_logging_decorator.gcpcxx.pb.cc, line 91 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5748#-MSjx7_BAN9abV5lWzSV:-MSjx7_BAN9abV5lWzSW:b-5uygos) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/3c51780ee9e5f0db333050bedf4e33776ff27e3e/generator/integration_tests/golden/internal/iam_credentials_logging_decorator.gcpcxx.pb.cc#L91)):*
> ```cpp
>   return google::cloud::internal::LogWrapper(
>       [this](grpc::ClientContext& context,
>              ::google::test::admin::database::v1::TailLogEntriesRequest const& request) ->
> ```

Consider:

```cc
[this](grpc::ClientContext& context, ::google::test::admin::database::v1::TailLogEntriesRequest const& request) {
  auto stream = child_->TailLogEntries(context, request);
  if (components_.count("rpc-streams") > 0) {
    stream =  absl::make_unique<internal::StreamingReadRpcLogging<
             ::google::test::admin::database::v1::TailLogEntriesResponse>>(std::move(stream), tracing_options_, internal::RequestIdForLogging());
  }
  return stream;
}
```

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5748)* status: 37 of 39 files reviewed, 4 unresolved discussions (waiting on @coryan and @scotthart)

___
*[generator/integration_tests/golden/iam_credentials_connection.gcpcxx.pb.cc, line 225 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5748#-MSjvVdW6ZdP6Wf9t8wm:-MSkQ0rj7LNRl1KmoL1j:b-2ca1jf) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/3c51780ee9e5f0db333050bedf4e33776ff27e3e/generator/integration_tests/golden/iam_credentials_connection.gcpcxx.pb.cc#L225)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Is this still needed?  I think you needed it to get the tracing parameters, but no longer use it?
</blockquote></details>

created #5813   
___
*[generator/integration_tests/golden/iam_credentials_metadata_decorator_test.cc, line 172 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5748#-MSjvyVP10fdZ3BQozd-:-MSkQdttEWumaUB0UPNT:b10nxfp) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/3c51780ee9e5f0db333050bedf4e33776ff27e3e/generator/integration_tests/golden/iam_credentials_metadata_decorator_test.cc#L172)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Should we write a templated mock and do this just once?  OTOH, it is hard to remember where these little mocks live.
</blockquote></details>

created issue #5814  
___
*[generator/integration_tests/golden/internal/iam_credentials_logging_decorator.gcpcxx.pb.cc, line 91 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5748#-MSjx7_BAN9abV5lWzSV:-MSkP7cz428_1LxflTz-:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/3c51780ee9e5f0db333050bedf4e33776ff27e3e/generator/integration_tests/golden/internal/iam_credentials_logging_decorator.gcpcxx.pb.cc#L91)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Consider:

```cc
[this](grpc::ClientContext& context, ::google::test::admin::database::v1::TailLogEntriesRequest const& request) {
  auto stream = child_->TailLogEntries(context, request);
  if (components_.count("rpc-streams") > 0) {
    stream =  absl::make_unique<internal::StreamingReadRpcLogging<
             ::google::test::admin::database::v1::TailLogEntriesResponse>>(std::move(stream), tracing_options_, internal::RequestIdForLogging());
  }
  return stream;
}
```

</blockquote></details>

Done.

**Comment by coryan** [comment]
___
*[generator/integration_tests/golden/iam_credentials_connection.gcpcxx.pb.cc, line 225 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5748#-MSjvVdW6ZdP6Wf9t8wm:-MSkUAXp3jeZbtNXdqx1:btvtdwj) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/3c51780ee9e5f0db333050bedf4e33776ff27e3e/generator/integration_tests/golden/iam_credentials_connection.gcpcxx.pb.cc#L225)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

created #5813   
</blockquote></details>

I do not follow:  the field and the constructor parameter is new **in this PR**, why add the field and a bug to maybe remove it later?  I think we should change the PR to not add the field.  If you do need it later you can add it when that motivation is clear.

**Review by coryan — CHANGES_REQUESTED** [review]
> Reviewed 1 of 40 files at r1, 4 of 4 files at r4.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5748)* status: all files reviewed, 4 unresolved discussions (waiting on @coryan and @scotthart)

___
*[generator/integration_tests/golden/iam_credentials_connection.gcpcxx.pb.cc, line 103 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5748#-MSmNF0ZEQKGb0_BPGum:-MSmNF0_3UwRyOYyIknE:b-h3z8ar) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/3c51780ee9e5f0db333050bedf4e33776ff27e3e/generator/integration_tests/golden/iam_credentials_connection.gcpcxx.pb.cc#L103)):*
> ```cpp
>       std::unique_ptr<BackoffPolicy> backoff_policy,
>       std::unique_ptr<IAMCredentialsConnectionIdempotencyPolicy> idempotency_policy,
>       IAMCredentialsConnectionOptions options)
> ```

This new parameter is not used, please remove it.
___
*[generator/integration_tests/golden/iam_credentials_connection.gcpcxx.pb.cc, line 112 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5748#-MSmNJGr-BI9TPS_MSpn:-MSmNJGs-J7a_OoH1J5H:brhumzt) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/3c51780ee9e5f0db333050bedf4e33776ff27e3e/generator/integration_tests/golden/iam_credentials_connection.gcpcxx.pb.cc#L112)):*
> ```cpp
>   explicit IAMCredentialsConnectionImpl(
>       std::shared_ptr<golden_internal::IAMCredentialsStub> stub,
>       IAMCredentialsConnectionOptions options)
> ```

Same here.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5748)* status: all files reviewed, 4 unresolved discussions (waiting on @coryan)

___
*[generator/integration_tests/golden/iam_credentials_connection.gcpcxx.pb.cc, line 103 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5748#-MSmNF0ZEQKGb0_BPGum:-MSnQjkNDIn2c061Gd31:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/3c51780ee9e5f0db333050bedf4e33776ff27e3e/generator/integration_tests/golden/iam_credentials_connection.gcpcxx.pb.cc#L103)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

This new parameter is not used, please remove it.
</blockquote></details>

Done.
___
*[generator/integration_tests/golden/iam_credentials_connection.gcpcxx.pb.cc, line 112 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5748#-MSmNJGr-BI9TPS_MSpn:-MSnQk2QAwCkQIeCPODm:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/3c51780ee9e5f0db333050bedf4e33776ff27e3e/generator/integration_tests/golden/iam_credentials_connection.gcpcxx.pb.cc#L112)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Same here.
</blockquote></details>

Done.
___
*[generator/integration_tests/golden/iam_credentials_connection.gcpcxx.pb.cc, line 225 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5748#-MSjvVdW6ZdP6Wf9t8wm:-MSnQMJvAMktBhq9xeeg:beoxez) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/3c51780ee9e5f0db333050bedf4e33776ff27e3e/generator/integration_tests/golden/iam_credentials_connection.gcpcxx.pb.cc#L225)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I do not follow:  the field and the constructor parameter is new **in this PR**, why add the field and a bug to maybe remove it later?  I think we should change the PR to not add the field.  If you do need it later you can add it when that motivation is clear.
</blockquote></details>

You are correct. After a few different implementations of this, I forgot which one I was on and thought the member field was still necessary.

**Review by coryan — APPROVED** [review]
:lgtm: maybe @devjgm also wants to take a look?

> Reviewed 4 of 7 files at r5.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5748#-:-MSnU0ta0kS5noH1RVkb:bm33pcz)* status: :shipit: complete! all files reviewed, all discussions resolved

**Pull request merged**